src/Playwright/Transport/Connection.cs

```diff
@@ -430,5 +420,89 @@ internal void TraceMessage(string logLevel, object message)
                 Console.Error.WriteLine(line);
             }
         }
+
+        internal async Task<T> WrapApiCallAsync<T>(Func<Task<T>> action, bool isInternal = false)
+        {
+            var st = new StackTrace(true);
+            var stack = new List<object>();
+            var lastInternalApiName = string.Empty;
+            var apiName = string.Empty;
+            for (int i = 0; i < st.FrameCount; ++i)
+            {
+                var sf = st.GetFrame(i);
+                string fileName = sf.GetFileName();
+                string cSharpNamespace = sf.GetMethod().ReflectedType.Namespace;
+                bool playwrightInternal = cSharpNamespace == "Microsoft.Playwright" || cSharpNamespace.StartsWith("Microsoft.Playwright.Core", StringComparison.InvariantCultureIgnoreCase) || cSharpNamespace.StartsWith("Microsoft.Playwright.Transport", StringComparison.InvariantCultureIgnoreCase);
+                if (string.IsNullOrEmpty(fileName) && !playwrightInternal)
+                    continue;
+
+                if (!playwrightInternal)
+                {
+                    stack.Add(new { file = fileName, line = sf.GetFileLineNumber(), column = sf.GetFileColumnNumber() });
+                }
+
+                string methodName = $"{sf?.GetMethod()?.DeclaringType?.Name}.{sf?.GetMethod()?.Name}";
+                if (methodName.Contains("WrapApiBoundaryAsync"))
+                {
+                    break;
+                }
+                if (methodName.StartsWith("<", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    continue;
+                }
+                if (playwrightInternal)
+                {
+                    lastInternalApiName = methodName;
+                }
+                else if (!string.IsNullOrEmpty(lastInternalApiName))
+                {
+                    apiName = lastInternalApiName;
+                    lastInternalApiName = string.Empty;
+                }
+            }
+            if (string.IsNullOrEmpty(apiName))
+            {
+                apiName = lastInternalApiName;
+            }
+            EnsureApiZoneExists();
+            if (ApiZone.Value[0] != null)
+            {
+                return await action().ConfigureAwait(false);
+            }
```

Comment:
Can this be moved to the top? It looks like the expensive `StackTrace(true)` stuff is only needed when there isn't an existing `ApiZone` already.